AO3-3828 Link to prompt in the notes of a prompt fill

bellisk opened the pull request:
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-3828

## Purpose

When a work has been submitted to fulfill a prompt in a Prompt Meme collection, an automatically-inserted text in the work notes says so. This PR adds a link to the specific prompt to that text.

## Testing Instructions

1. Post a work that fulfills a prompt
2. Look at the automatically-inserted text in the work notes
3. Note that it says: "In response to a prompt by CREATOR in the COLLECTION_NAME collection."
4. Note that the word "prompt" is a link to the prompt in question.

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?

katzenfabrik, they/them

bellisk commented:
The failing cucumber test passes locally (in Docker) and I really can't think of a way my change could affect admin caching of public skins.

sarken reviewed (changes requested):
Thanks for this! However, I think it would be a good idea to extend the tests a bit and ensure that following the "prompt" link works, especially if the person following the link is logged out or logged in as a random user (i.e., not the challenge mod, work creator, or prompter).

bellisk commented:
@sarken Thanks, you're right. I'll add more tests.

bellisk commented:
Thanks @sarken - once I added those tests, I realised that logged-out users didn't have permission to see individual prompts, so I've fixed that too now. :)

sarken commented on features/prompt_memes_b/challenge_promptmeme_posting_fills.feature:
```diff
@@ -458,3 +459,41 @@ Feature: Prompt Meme Challenge
+  Scenario: Fic links to the prompt it fulfils, for all users
+
+  Given I have Battle 12 prompt meme fully set up
+  When I am logged in as "myname1"
+  When I sign up for Battle 12 with combination B
+    And I am logged in as "myname4"
+    And I claim a prompt from "Battle 12"
+  When I fulfill my claim
+  When I reveal works for "Battle 12"
```
Let's make all of these start with "And" and be indented under the initial "Given" step, since they're all sort of background/setup steps.

sarken commented on features/prompt_memes_b/challenge_promptmeme_posting_fills.feature:
```diff
@@ -458,3 +459,41 @@ Feature: Prompt Meme Challenge
+  Then I should see "Fulfilled Story"
+    And I should see "In response to a prompt by Anonymous"
+    And I should see a link "prompt"
+    And I follow "prompt"
+    And I should see "Request by Anonymous"
```
When we're switching from checking the results (all of those "I should see" steps) to taking an action, we like to switch back to "When" -- and then back to "Then" for checking the results. So:

```
When I follow "prompt"
Then I should see "Request by Anonymous"
```

Lines 484 & 485, 491 & 492, and 498 & 499 also need this change.

sarken reviewed (changes requested):
Yay for tests catching problems! I just have a couple of minor style suggestions about the tests, and then I think this will be ready to merge 👍

bellisk commented:
@sarken Thanks for the codestyle tips! It's a lot clearer to me now how those cucumber tests are supposed to work. I've made the fixes you suggested. :)

sarken reviewed (approved):
This slipped through the cracks on me -- my apologies for that, and thanks for being patient! It looks good, so I'm going to stick it in our next release.
